azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,277 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from multiprocessing.sharedctypes import Value
+from knack.log import get_logger
+from azext_iot.operations import hub
+from azext_iot.sdk.iothub import device
+from azure.cli.core.azclierror import (
+    CLIInternalError,
+    InvalidArgumentValueError,
+    RequiredArgumentMissingError,
+)
+from azext_iot.common.shared import SdkType, JobStatusType, JobType, JobVersionType
+from azext_iot.common.utility import handle_service_exception, process_json_arg
+from azext_iot.operations.generic import _execute_query, _process_top
+from azext_iot.iothub.providers.base import IoTHubProvider, CloudError, SerializationError
+from azext_iot.operations.hub import _iot_device_show, _iot_device_twin_show, _iot_device_module_twin_show, _iot_device_module_show
+from azext_iot.operations.hub import iot_device_list, iot_device_create, iot_device_delete, iot_device_twin_replace, iot_hub_configuration_create, iot_hub_configuration_list, \
+    iot_hub_configuration_delete, iot_device_module_list, iot_device_module_create, iot_device_module_twin_replace
+
+import json
+
+logger = get_logger(__name__)
+
+
+class StateProvider(IoTHubProvider):
+    def __init__(self, cmd, hub, rg, orig_hub=None, orig_rg=None, auth_type_dataplane=None):
+        super().__init__(
+            cmd=cmd, 
+            hub_name=hub,
+            rg=rg, 
+            auth_type_dataplane=auth_type_dataplane
+        )
+        self.auth_type = auth_type_dataplane
+        self.orig_hub = orig_hub
+        self.orig_rg = orig_rg
+
+        if(orig_hub):
+            self.orig_hub_target = self.discovery.get_target(
+                resource_name=self.orig_hub,
+                resource_group_name=self.orig_rg,
+                auth_type=auth_type_dataplane,
+            )
+
+    def save_state(self, filename):
+        '''
+        Writes all hub configurations, device identities and device twins from the origin hub to a json file
+        '''
+
+        configs = iot_hub_configuration_list(cmd=self.cmd, hub_name=self.hub_name, resource_group_name=self.rg, auth_type_dataplane=self.auth_type)
+
+        identities = iot_device_list(cmd=self.cmd, hub_name=self.hub_name, resource_group_name=self.rg, auth_type_dataplane=self.auth_type)
+
+        with open(filename, 'w') as f:
+
+            json.dump(configs, f)
+            f.write('\n')
+
+            for id in identities: 
+
+                module_objs = iot_device_module_list(cmd=self.cmd, device_id=id["deviceId"], hub_name=self.hub_name, resource_group_name=self.rg, auth_type_dataplane=self.auth_type)
+
+                if(id["capabilities"]["iotEdge"]):
+                    id["numModules"] = len(module_objs) - 2
```

Comment:
Why are we doing a -2 here? It will be good to add some comments across the code.